### PR TITLE
fix: only run requestIdleCallback if it exists

### PR DESCRIPTION
**Pull request by raunofreiberg**
Currently, the app would crash in Safari because it [does not support `requestIdleCallback`](https://caniuse.com/#search=requestIdleCallback)

Maybe we should polyfill it or use an alternative if performance becomes an issue.

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -195,21 +195,28 @@ export default function AxeMode({ children, disabled }: AxeModeProps) {
     }
 
     if (childrenRef.current) {
-      if (idleId) {
+      if (idleId && 'cancelIdleCallback' in window) {
```
Comment (chancestrickland) [inline src/index.tsx]: I think this is a good approach with `requestIdleCallback`. I think bundling a polyfill for an experimental API is a bit much for a relatively minor optimization.

File: src/index.tsx [hunk 2]
```diff
@@ -195,21 +195,28 @@ export default function AxeMode({ children, disabled }: AxeModeProps) {
...
         // requestIdleCallback has no types:
         // https://github.com/microsoft/TypeScript/issues/21309
         // @ts-ignore
         cancelIdleCallback(idleId);
         setIdleId(null);
```
Comment (chancestrickland) [inline src/index.tsx]: I'd probably opt for a ref here vs. state. No reason to re-render just to capture the ID. Same comment applies elsewhere. 

`idleId.current = null`

File: src/index.tsx [hunk 3]
```diff
@@ -195,21 +195,28 @@ export default function AxeMode({ children, disabled }: AxeModeProps) {
...
       }
 
-      // requestIdleCallback has no types:
-      // https://github.com/microsoft/TypeScript/issues/21309
-      // @ts-ignore
-      const id = requestIdleCallback(() => {
+      function getViolations() {
```
Comment (chancestrickland) [inline src/index.tsx]: In strict mode the compiler will get angry about this function definition being inside of the `if` block. I'd move it outside the block but still in the effect.

File: src/index.tsx [hunk 4]
```diff
@@ -195,21 +195,28 @@ export default function AxeMode({ children, disabled }: AxeModeProps) {
...
+        // requestIdleCallback has no types:
+        // https://github.com/microsoft/TypeScript/issues/21309
+        // @ts-ignore
+        const id = requestIdleCallback(getViolations);
+        setIdleId(id);
+      } else {
+        getViolations();
+      }
     }
   }, [children, disabled]);
```
Comment (chancestrickland) [inline src/index.tsx:219]: This won't reliably update when the DOM updates, just when `children` changes. The DOM could be updated by a queued async task, for example and our component may not capture violations in the updated node tree. I might remove the `children` dependency and use a `MutationObserver` to react to actual DOM changes

File: src/index.tsx (hunk 2 above)
Comment (chancestrickland) [reply src/index.tsx]: NOTE: I went ahead and pushed up some of these changes, but feel free to review and push back and/or revert if you think I missed the intent here.

File: example/index.tsx [hunk 5]
```diff
@@ -29,7 +29,10 @@ const App = () => {
           <h6>Accessibility testing componentized</h6>
           <input placeholder="Username" />
           <input placeholder="Password" />
-          <button aria-expanded="foo" onClick={() => console.log('yo')}>
+          <button
+            aria-expanded={'foo' as any}
```
Comment (chancestrickland) [inline example/index.tsx:33]: Just to make TS happy since this is intentionally invalid

File: src/index.tsx (hunk 2 above)
Comment (raunofreiberg) [reply src/index.tsx]: Thanks for that! Makes sense to me.

File: src/index.tsx (hunk 4 above)
Comment (raunofreiberg) [reply src/index.tsx:219]: That's true. It also wouldn't catch any imperative DOM mutations done in effects, right? Such as the ones we do to place the overlays.

Maybe let's test this out in real apps and do it as a follow-up? 🙂

**Pull request merged**